### PR TITLE
1.6/1329 add centre backend mechanics

### PR DESCRIPTION
**Pull request by charlesstrange2**
review this AFTER
https://github.com/neontribe/ARCVService/pull/285
It's dependent on it.

### CODE REVIEW

File: app/Rules/NotExistsRule.php [hunk 1]
```diff
@@ -37,7 +37,7 @@ public function passes($attribute, $value)
             : $this->parameters[1];
 
         return DB::table($tableName)
-            ->where($columnName, $value)
+            ->where($columnName, 'like', $value)
```
Comment (katjam) [inline app/Rules/NotExistsRule.php]: Do you need to surround the $value with %%

File: resources/views/service/centres/create.blade.php [hunk 2]
```diff
@@ -31,12 +30,13 @@
                     </select>
                 </div>
                 <div class="select">
-                    <label for="form">Form</label>
-                    <select name="form" id="form" class="{{ $errors->has('form') ? 'error' : '' }}" required>
-                        <option value="">Choose one</option>
+                    <label for="print_pref">Printed Form</label>
+                    <select name="print_pref" id="print_pref" class="{{ $errors->has('print_pref') ? 'error' : '' }}" required>
+                        <option value="" disabled selected>Choose one</option>
```
Comment (katjam) [inline resources/views/service/centres/create.blade.php]: Text outside the quotes

File: tests/Unit/Controllers/Service/Admin/CentreControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    public function setUp()
+    {
+        parent::setUp();
+        $this->faker = Factory::create(config('app.locale'));
+        $this->adminUser = factory(AdminUser::class)->create();
+        $this->sponsor = factory(Sponsor::class)->create();
+        $this->data = [
+            'name' => $this->faker->city,
+            'sponsor' => $this->sponsor->id,
+            'rvid_prefix' => strtoupper($this->faker->lexify(str_repeat('?', rand(1, 5)))),
```
Comment (katjam) [inline tests/Unit/Controllers/Service/Admin/CentreControllerTest.php]: Neat trick.

File: tests/Unit/FormRequests/AdminNewCentreRequestTest.php [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        );
+    }
+
+    /**
+     * must return hardcoded values
+     * @return array
+     */
+    public function storeValidationProvider()
+    {
+        // We need to do this as this provider gets made before laravel normally gets out of bed.
```
Comment (katjam) [inline tests/Unit/FormRequests/AdminNewCentreRequestTest.php]: :grin:

File: app/Rules/NotExistsRule.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Rules/NotExistsRule.php]: it's a cheat, i'm doing it to deal with uppercase/lowercase things.

Comment (charlesstrange2) [reply app/Rules/NotExistsRule.php]: i'll comment that.

File: resources/views/service/centres/create.blade.php (hunk 2 above)
Comment (charlesstrange2) [reply resources/views/service/centres/create.blade.php]: yep, they're the null attributes `disabled` and `selected`, so users must  choose a valid one and and it's not picked a default.

File: tests/Unit/Controllers/Service/Admin/CentreControllerTest.php (hunk 3 above)
Comment (charlesstrange2) [reply tests/Unit/Controllers/Service/Admin/CentreControllerTest.php]: :+1:

File: app/Rules/NotExistsRule.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Rules/NotExistsRule.php]: scratch that, changed it.

File: resources/views/service/centres/create.blade.php (hunk 2 above)
Comment (katjam) [reply resources/views/service/centres/create.blade.php]: :man_facepalming: Of course!

**Pull request merged**